flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/optimize/program/FlinkRuntimeFilterProgramTest.java

```diff
@@ -0,0 +1,271 @@
+        util.verifyRelPlan(query);
+        util.verifyExecPlan(query);
+    }
+
+    @Test
+    public void testProbeSideIsTooSmall() throws Exception {
+        // runtime filter will not succeed
+        setupTableRowCount("dim", SUITABLE_DIM_ROW_COUNT);
+        // fact is 7.5 GB < 10 GB
+        setupTableRowCount("fact", 128L * 1024L * 1024L);
+        String query = "select * from fact, dim where fact.amount = dim.amount and dim.price < 500";
+        util.verifyRelPlan(query);
+        util.verifyExecPlan(query);
+    }
+
+    @Test
+    public void testBuildSideIsTooLarge() throws Exception {
+        // runtime filter will not succeed
+        // dim is 48 MB > 6MB
+        setupTableRowCount("dim", 1024L * 1024L);
+        setupTableRowCount("fact", SUITABLE_FACT_ROW_COUNT);
+        String query = "select * from fact, dim where fact.amount = dim.amount and dim.price < 500";
+        util.verifyRelPlan(query);
+        util.verifyExecPlan(query);
+    }
+
+    @Test
+    public void testFilterRatioIsTooSmall() throws Exception {
+        // runtime filter will not succeed
+        setupSuitableTableStatistics();
+        setupTableColumnNdv("dim", "amount", 768L);
+        setupTableColumnNdv("fact", "amount", 1024L);
+        String query = "select * from fact, dim where fact.amount = dim.amount and dim.price < 500";
+        util.verifyRelPlan(query);
+        util.verifyExecPlan(query);
+    }
+
+    @Test
+    public void testBuildSideIsJoinWithoutExchange() throws Exception {
+        setupSuitableTableStatistics();
+        util.tableEnv()
+                .executeSql(
+                        "create table fact2 (\n"
+                                + "  id BIGINT,\n"
+                                + "  amount BIGINT,\n"
+                                + "  price BIGINT\n"
+                                + ") with (\n"
+                                + " 'connector' = 'values',\n"
+                                + "  'runtime-source' = 'NewSource',\n"
+                                + " 'bounded' = 'true'\n"
+                                + ")");
+        setupTableRowCount("fact2", SUITABLE_FACT_ROW_COUNT);
+
+        String query =
+                "select * from dim, fact, fact2 where fact.amount = fact2.amount and"
+                        + " fact.amount = dim.amount and dim.price < 500";
+        util.verifyRelPlan(query);
+        util.verifyExecPlan(query);
+    }
+
+    @Test
+    public void testBuildSideIsAggWithoutExchange() throws Exception {
+        setupSuitableTableStatistics();
+        String query =
+                "select * from fact join"
+                        + " (select dim_date_sk, sum(dim.price) from dim where"
+                        + "  dim.price < 500 group by dim_date_sk) dimSide"
+                        + " on fact.fact_date_sk = dimSide.dim_date_sk";
+        util.verifyRelPlan(query);
+        util.verifyExecPlan(query);
+    }
+
+    @Test
```

Comment:
I have one question: do we need to consider conflicts between runtime filters and DPP for partitioned tables?

Reply:
I think the runtime filter doest not conflict with DPP, they can exist at the same time. WDYT ?

Reply:
> I think the runtime filter doest not conflict with DPP, they can exist at the same time. WDYT ?

Yes, I understand now